beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+// if it finds a saved state the the tail block was descended from, it returns this state and
+// all blocks in the lineage, including the tail block. blocks are returned in ascending order.
+// note that this function assumes that the tail is a canonical block, and therefore assumes that
+// all ancestors are also canonical.
+func (r *CanonicalChainer) ancestorChain(ctx context.Context, tail block.SignedBeaconBlock) (state.BeaconState, []block.SignedBeaconBlock, error) {
+	ctx, span := trace.StartSpan(ctx, "CanonicalChainer.ancestorChain")
+	defer span.End()
+	head := &link{b: tail}
+	for {
+		if err := ctx.Err(); err != nil {
+			msg := fmt.Sprintf("context canceled while finding ancestors of block @ slot %d", tail.Block().Slot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		b := head.b.Block()
+		// compute hash_tree_root of current block and try to look up the corresponding state
+		root, err := b.HashTreeRoot()
+		if err != nil {
+			msg := fmt.Sprintf("could not compute htr for descendent block @ slot=%d", b.Slot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		st, err := r.h.StateOrError(ctx, root)
+		// err from StateOrError == nil, we've got a real state - the job is done!
+		if err == nil {
+			// this handles cases where we find the state right away. return empty list of blocks
+			if head.child == nil {
+				return st, []block.SignedBeaconBlock{}, nil
+			}
+			// we found the state by the root of the head, meaning it has already been applied.
+			// we only want to return the blocks descended from it.
+			return st, head.child.toSlice(), nil
+		}
+		// ErrStateNotFound errors are fine, but other errors mean something is wrong with the db
+		if !errors.Is(err, dberr.ErrStateNotFound) {
+			return nil, nil, errors.Wrap(err, fmt.Sprintf("error querying database for state w/ block root = %#x", root))
+		}
+		parent, err := r.h.Block(ctx, bytesutil.ToBytes32(b.ParentRoot()))
+		if err != nil {
+			msg := fmt.Sprintf("db error when retrieving parent of block @ slot=%d by root=%#x", b.Slot(), b.ParentRoot())
+			return nil, nil, errors.Wrap(err, msg)
+		}
+		if helpers.BeaconBlockIsNil(parent) != nil {
+			msg := fmt.Sprintf("unable to retrieve parent of block @ slot=%d by root=%#x", b.Slot(), b.ParentRoot())
+			return nil, nil, errors.Wrap(dberr.ErrNotFound, msg)
+		}
+		head = head.prepend(parent)
+	}
+}
+
+// link is a helper type for reversing the chain
+type link struct {
+	b           block.SignedBeaconBlock
+	child       *link
+	numChildren int
+}
+
+func (head *link) toSlice() []block.SignedBeaconBlock {
+	blocks := make([]block.SignedBeaconBlock, head.numChildren+1)
+	for i := 0; i < len(blocks); i++ {
+		blocks[i] = head.b
+		head = head.child
+	}
+	return blocks
+}
+
+func (head *link) prepend(b block.SignedBeaconBlock) *link {
+	return &link{
+		b:           b,
+		child:       head,
+		numChildren: head.numChildren + 1,
+	}
+}
```

Comment:
Maybe we can put the chaniner stuff in `chainer.go`. There's a lot going on in this file

Reply:
It also feels more neutral to export these in db pkg rather than stategen pkg if we want to expose it

Reply:
agreed, making it private